### ISSUE TITLE
Possible arbitrary code execution

### ISSUE THREAD

**Issue by Yoshino-s** [donmccurdy/expression-eval#52]
Althrough we have decleared in the README.MD that do not use this package with user-provided inputs, but after i exam some project with this project, i found that many developers still use in that way, which may lead to some serious security problem. So I think that we still need to do some basic thing to prevent some dangerous action in the package itself.

I have made a tiny fix to prevent access prototype on object, which may prevent all prototype pollution and code execution.

https://github.com/418sec/expression-eval/pull/1

Should we use this fix or just ignore the possible vuln?

**Comment by donmccurdy** [comment]
Hi @Yoshino-s! Thanks for looking into this.

I believe you're correct that the fix in https://github.com/418sec/expression-eval/pull/1 will prevent the attack mentioned in this project's README. However, that is only one example of a class of possible vulnerabilities — for example, passing a `Function` constructor or any reference to the global scope will allow similar attacks. I'm not able to identify all possible attack vectors, so the risk of user-provided inputs remains.

Perhaps the specific vulnerability here is worth fixing, even though we can't guarantee safety with user-provided expressions and inputs. I can't immediately think of another exploit that works when all inputs are primitive values, but they may exist. I would certainly be curious to know if there are suggested resources on this type of issue, or other ideas.

**Comment by Yoshino-s** [comment]
Indeed, if we directly pass an Function into context, we have no simple solution to prevent the attack. There is an complex solution like `vm2` or other vm packages but that's not what we want.

We should only prevent attack like 

```js
const expression = req.expression; // use4 input.
const res = eval(parse(expression), {
  x: 1, y: 2
})
```

Also this fix prevent all access to prototype chain, so we can safely pass an object or others into context.

The possible attack still exist only when passing an dangerous object or function into context, like `eval(expr, {process})`, which i think no one will use like this.

**Comment by Yoshino-s** [comment]
By the way, should we submit it to github security advisory and npm advisory, which will automatically alert downstream package and app? And, can we apply for an CVE ID for the vuln, which can help me a lot?
Thanks a lot.

**Comment by donmccurdy** [comment]
> The possible attack still exist only when passing a dangerous object or function into context, like eval(expr, {process}), which i think no one will use like this.

I don't feel comfortable assuming this. The attack surface of all JavaScript objects is too wide, and there are any number of subtle ways a developer could accidentally expose access to a constructor or another eval-like interface. For example, `Reflect.get(String, 'constructor')`. 

In short, the warning appearing on the README probably cannot be removed — users need to be aware that untrusted inputs may alter application state. In a client-side application, where the user can access the JS Console anyway, this is completely fine. In a server-side application, if anyone can send expressions to the application, it would represent a vulnerability in the server application. I don't believe that this can be considered a vulnerability in this library, though, in the same way that an HTML rendering library is not "vulnerable" when a developer chooses to put unsafe input into un-escaped HTML.

Perhaps it is worth mentioning `vm2` in the readme, for users who need that level of security in a Node.js environment.

**Comment by Yoshino-s** [comment]
I agree with you that the warning in README shouldn't be removed. What I think is that we shouldn't add more attack possibility inside an package.

Passing `{Function}` or other dangerous things by developers should be developers' responsibility to keep input safe, because they should aware that this object will lead to an unexpectable result, and also that shouldn't be the correct purpose of this package. This package is only used to do some mathmatical calculation.

But if developers use this like `eval(expr, {x:1})` which they think totally correct, we should try to keep no possible attack to it, also it's the main purpose of this package.

In short, the fix is necessary and should be fixed, but we should keep warning users.

Also, i would spend some time on downstream project and try to fix.

**Comment by Yoshino-s** [comment]
any progress here?

**Comment by donmccurdy** [comment]
Thanks @Yoshino-s, after some discussion with `huntr.dev` support I've gone ahead with merging the fix. To clarify for future readers, this fix (to be released in v5.0) provides valuable — but still incomplete — protection against a specific class of user inputs. Evaluation with arbitrary user inputs remains unsafe.

Safely evaluating arbitrary user input is still out of scope for this small library, as mentioned in the readme, and so any downstream projects using the library with other expectations should probably be considering something like `vm2` instead.
